### ISSUE TITLE
Problem with the *experimental* call by need

### ISSUE THREAD

**Issue by teofr** [astampoulis/makam#52]
Under the `experimental/examples` there's an issue with the call by need strategy.

In particular, in the evaluation rule for the `app` we introduce a local variable `thunk` that we'll use to simulate the call by name part of call by need, and a new clause stating how to evaluate this variable. Using `memoized` to implement sharing.

https://github.com/astampoulis/makam/blob/f97a3ba69556f56b2967cac89d41801c6a24f097/examples/experiments/stlc_strategies.makam#L41-L46

The problem is that the value V might be a lambda abstraction, with appearings of the new variable. In this case we would need to provide a global variable instead of a local one. I'm not completely sure if there's an idiomatic way to do this on Makam (or if it's even possible).

**Example of failure:**

If we execute the code as is, evaluating it prints:
```
-- Query result in file ./examples/experiments/stlc_strategies.makam, line 63, character 2 to line 66, character 71:
cbv eval
"evaluation"
tuple [ (tuple [  ]), (tuple [  ]) ]
cbname eval
"evaluation"
"evaluation"
tuple [ (tuple [  ]), (tuple [  ]) ]
cbneed eval
"evaluation"
tuple [ (tuple [  ]), (tuple [  ]) ]

```

If we change the following part ... 

https://github.com/astampoulis/makam/blob/f97a3ba69556f56b2967cac89d41801c6a24f097/examples/experiments/stlc_strategies.makam#L63-L66

for:

```
(eq TERM (app (app (lam (fun x => lam (fun y => tuple [x, x, y, y])))  side_effect) (tuple [])),
 print_string "cbv eval\n", eval_cbv TERM CBV_VAL, print CBV_VAL,
 print_string "cbname eval\n", eval_cbnm TERM CBNM_VAL, print CBNM_VAL,
 print_string "cbneed eval\n", eval_cbnd TERM CBND_VAL, print CBND_VAL) ?
```

Evaluating it prints:

```
-- Query result in file ./examples/experiments/stlc_strategies.makam, line 63, character 2 to line 66, character 71:
cbv eval
"evaluation"
tuple [ (tuple [  ]), (tuple [  ]), (tuple [  ]), (tuple [  ]) ]
cbname eval
"evaluation"
"evaluation"
tuple [ (tuple [  ]), (tuple [  ]), (tuple [  ]), (tuple [  ]) ]
cbneed eval
Impossible.
```

**Comment by teofr** [comment]
As another example, if we change those last lines to

```
(eq TERM (app (app (lam (fun x => lam (fun y => tuple [y, y])))  side_effect) (tuple [])),
 print_string "cbv eval\n", eval_cbv TERM CBV_VAL, print CBV_VAL,
 print_string "cbname eval\n", eval_cbnm TERM CBNM_VAL, print CBNM_VAL,
 print_string "cbneed eval\n", eval_cbnd TERM CBND_VAL, print CBND_VAL) ?
```

taking the variable bound by the outer lambda from the body, we no longer need to unify `thunk` with the outside world during the evaluation of the inner `app`, and evaluating prints:

```
-- Query result in file ./examples/experiments/stlc_strategies.makam, line 62, character 2 to line 65, character 71:
cbv eval
"evaluation"
tuple [ (tuple [  ]), (tuple [  ]) ]
cbname eval
tuple [ (tuple [  ]), (tuple [  ]) ]
cbneed eval
tuple [ (tuple [  ]), (tuple [  ]) ]
```

**Comment by astampoulis** [comment]
Hi @teofr, thanks for reporting the issue. Indeed the strategy that the experiment implements does not work properly in the presence of lambdas -- and your first counterexample is a good example of that. I am not sure if I understand your second counterexample -- isn't this the behavior you would expect for this example?

The closest think to global variables in Makam are unification variables, which are indeed used here as a trick to get this to work. (`V2` starts out as an uninstantiated unification variable, then once the evaluation result of `E2` is known, `V2` becomes instantiated with it, and then the result gets re-used  instead of re-evaluating `E2`. The fact that this trick 'works' is what this experiment was initially about.)  I'll push a PR shortly that eschews the local `thunk` variable in favor of a `thunk E2 V2` term constructor, and has the expected behavior for your counterexample.

I am thinking also of adding an example that models a heap explicitly in the semantics, and implements the standard Launchbury semantics for call-by-need, without Makam-specific tricks.

**Comment by teofr** [comment]
Thanks for the reply!

> I am not sure if I understand your second counterexample -- isn't this the behavior you would expect for this example?

Yeah, sorry, I wasn't really clear, this example just helped me convince myself of what was going on.

> I am thinking also of adding an example that models a heap explicitly in the semantics, and implements the standard Launchbury semantics for call-by-need, without Makam-specific tricks.

This would be pretty cool.

**Comment by astampoulis** [comment]
> > I am thinking also of adding an example that models a heap explicitly in the semantics, and implements the standard Launchbury semantics for call-by-need, without Makam-specific tricks.
> This would be pretty cool.

Sounds good, I'm hoping I'll have a PR for that in the next few days.

**Comment by teofr** [comment]
Closed by #54
